web/models/key.ex

```diff
@@ -31,11 +33,35 @@ defmodule HexWeb.Key do
   end
 
   def all(user) do
-    assoc(user, :keys)
+    from(k in assoc(user, :keys), where: is_nil(k.revoked_at))
   end
 
   def get(name, user) do
-    from(k in assoc(user, :keys), where: k.name == ^name)
+    from(k in assoc(user, :keys), where: k.name == ^name and is_nil(k.revoked_at))
+  end
+
+  def get_revoked(name, user) do
+    from(k in assoc(user, :keys), where: k.revoked_name == ^name)
+  end
+
+  def revoke(user, key_name, revoked_at \\ Ecto.DateTime.utc) do
+    from(k in assoc(user, :keys),
+      where: k.name == ^key_name and is_nil(k.revoked_at),
+      update: [set: [
+        revoked_name: k.name,
+        revoked_at: fragment("?", ^revoked_at),
+        name: nil
+      ]])
+  end
```

Comment:
Could we use a changeset here instead? Or are there instances where we need to revoke a key that we haven't already fetched?